### PR TITLE
Change `trainer.should_stop` to not stop in between an epoch and run until `min_steps/min_epochs` only

### LINKED ISSUE #13789: `trainer.should_stop` doesn't stop at `min_epochs` if it's set before
## Proposed refactor

Consider an example:
```
min_epochs=4
max_epochs=1000

user set's `should_stop=True` in the middle of epoch 3
```

Result:
```
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> (epoch 1)
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> (epoch 2)
>>>>>>>>>>>>>>> (should_stop=True, epoch 3 breaks)
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> (epoch 4)
...
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> (max_epochs)
```
1. The epoch loop breaks in between at epoch 3.
2. Training continues until `max_epochs=1000`.

### Motivation and Pitch

My suggestion:
1. Training should not break in between since `min_epochs` haven't been met.
2. It should break right after `min_epochs`.

Motivation or Reason
1. Breaking the loop in between is fine if `min_epochs` is met, but if it's not, it should not break it in between and should complete that epoch and continue until `min_epochs`.
2. If the user has been forced to stop the training, we should not run for more than `min_epochs`.

Thoughts??

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/Lightning-AI/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/Lightning-AI/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/Lightning-AI/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/Lightning-AI/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @tchaton @justusschock @awaelchli @borda @carmocca @ananthsub @ninginthecloud @rohitgr7

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #13789

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @justusschock @awaelchli @rohitgr7 @borda @carmocca @ananthsub @ninginthecloud @akihironitta

### CODE REVIEW

**Review by awaelchli — COMMENTED** [review]
Initial comments, currently studying the loop changes and tests

File: src/pytorch_lightning/CHANGELOG.md [hunk 1]
```diff
@@ -170,6 +170,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Disallowed using `BatchSampler` when running on multiple IPUs ([#13854](https://github.com/PyTorchLightning/pytorch-lightning/pull/13854))
 
 
+- Refactored `trainer.should_stop` to not stop in between an epoch and run until `min_steps/min_epochs` only ([#13890](https://github.com/Lightning-AI/lightning/pull/13890))
+
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply src/pytorch_lightning/CHANGELOG.md:174]: It's a behavior change.. will put this under `Changes`

File: docs/source-pytorch/common/trainer.rst [hunk 2]
```diff
@@ -1761,3 +1761,33 @@ execution within that function, and the status of the Trainer.
     trainer.state.status
     # stage in ("train", "sanity_check", "validate", "test", "predict", "tune")
     trainer.state.stage
+
+should_stop
+***********
+
+If you want to terminate the training during ``.fit``, you can set ``trainer.should_stop=True`` and Lightning will terminate the training
+immediately. Note that, it will respect the arguments ``min_steps`` and ``min_epochs`` to check whether to stop it or not. If these
```
[the remark this answers is not in the document]
Comment (carmocca) [reply docs/source-pytorch/common/trainer.rst:1769]: I agree. `immediately` would be something like raising an exception to stop right after the attribute is set.

File: docs/source-pytorch/common/trainer.rst [hunk 3]
```diff
@@ -1761,3 +1761,33 @@ execution within that function, and the status of the Trainer.
...
+arguments are set and the ``current_epoch`` or ``global_step`` doesn't meet these minimum conditions, training will continue until
+both conditions are met. If any of these arguments is not set, it won't be considered for the final decision.
+
+
+.. code-block:: python
+
+    trainer = Trainer()
+    # setting `trainer.should_stop`` at any point of training will terminate
+    trainer.should_stop = True
```
Comment (carmocca) [inline docs/source-pytorch/common/trainer.rst:1778]: What usecase is this example trying to demostrate? Nobody would ever do this

I think this example needs to show that `should_stop` is set at any point DURING training (so in a hook somewhere)

File: src/pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 4]
```diff
@@ -102,7 +102,23 @@ def _is_validation_done(self) -> bool:
...
     def done(self) -> bool:
         """Evaluates when to leave the loop."""
-        return (self._is_training_done and self._is_validation_done) or self.trainer.should_stop
+        if self._is_training_done and self._is_validation_done:
+            return True
+
+        if self.trainer.should_stop:
+            # early stopping
+            min_epochs = self.trainer.fit_loop.min_epochs
+            met_min_epochs = self.trainer.current_epoch >= min_epochs if min_epochs else True
```
Comment (carmocca) [inline src/pytorch_lightning/loops/epoch/training_epoch_loop.py:111]: I would create protected propertiees for these checks, so we reuse the code in both loops

File: src/pytorch_lightning/loops/fit_loop.py [hunk 5]
```diff
@@ -174,15 +174,9 @@ def done(self) -> bool:
             met_min_epochs = self.epoch_progress.current.processed >= self.min_epochs if self.min_epochs else True
             met_min_steps = self.epoch_loop.global_step >= self.min_steps if self.min_steps else True
             if met_min_epochs and met_min_steps:
-                self.trainer.should_stop = True
```
Comment (carmocca) [inline src/pytorch_lightning/loops/fit_loop.py:177]: Shouldn't this stay?

`should_stop = met_min_epochs and met_min_steps`

File: tests/tests_pytorch/trainer/test_trainer.py [hunk 6]
```diff
@@ -599,7 +599,10 @@ def training_step(self, batch, batch_idx):
             output["loss"] = output["loss"] * 0.0  # force minimal loss to trigger early stopping
             self.log("loss", output["loss"])
             self.training_step_invoked += 1
-            assert not self.trainer.should_stop
+            if self.current_epoch < 2:
+                assert not self.trainer.should_stop
+            else:
+                assert self.trainer.should_stop
```
Comment (carmocca) [inline tests/tests_pytorch/trainer/test_trainer.py:605]: Why is this ever True?

File: tests/tests_pytorch/callbacks/test_early_stopping.py [hunk 7]
```diff
@@ -265,25 +265,28 @@ def validation_epoch_end(self, outputs):
...
-@pytest.mark.parametrize("limit_train_batches", (3, 5))
 @pytest.mark.parametrize(
-    ["min_epochs", "min_steps"],
+    "limit_train_batches,min_epochs,min_steps,stop_step",
     [
         # IF `min_steps` was set to a higher value than the `trainer.global_step` when `early_stopping` is being
         # triggered, THEN the trainer should continue until reaching `trainer.global_step == min_steps` and stop
-        (0, 10),
+        # (3, 0, 10, 10),
+        # (5, 0, 10, 10),
```
Comment (carmocca) [inline tests/tests_pytorch/callbacks/test_early_stopping.py:274]: Commented because it's WIP?

File: src/pytorch_lightning/loops/fit_loop.py (hunk 5 above)
Comment (rohitgr7) [reply src/pytorch_lightning/loops/fit_loop.py:177]: it's already `True`. L-172

File: src/pytorch_lightning/utilities/warnings.py [hunk 8]
```diff
@@ -39,6 +40,11 @@ def deprecation(self, message: str, stacklevel: int = 5, **kwargs: Any) -> None:
             self.add(message)
             new_rank_zero_deprecation(message, stacklevel=stacklevel, **kwargs)
 
+    def info(self, message: str, stacklevel: int = 5, **kwargs: Any) -> None:
```
Comment (rohitgr7) [inline src/pytorch_lightning/utilities/warnings.py:43]: we can rename WarningCache to `LoggingCache` if required since `info` is not a warning.

File: tests/tests_pytorch/callbacks/test_early_stopping.py (hunk 7 above)
Comment (rohitgr7) [reply tests/tests_pytorch/callbacks/test_early_stopping.py:274]: ah! no.. uncommented.

File: tests/tests_pytorch/trainer/test_trainer.py (hunk 6 above)
Comment (rohitgr7) [reply tests/tests_pytorch/trainer/test_trainer.py:605]: because min_epochs is 5, `should_stop` will remain `True`for some time until `min_epochs` are completed.

patience is 0 here. After 1st epoch, it will store the loss. In the 2nd epoch, it will recheck it and trigger `should_stop=True`. And from 3rd to `min_epochs` it will remain `True`and will terminate after the 5th epoch.

File: docs/source-pytorch/common/trainer.rst (hunk 3 above)
Comment (rohitgr7) [reply docs/source-pytorch/common/trainer.rst:1778]: updated!

File: src/pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 4 above)
Comment (rohitgr7) [reply src/pytorch_lightning/loops/epoch/training_epoch_loop.py:111]: done!

File: tests/tests_pytorch/callbacks/test_early_stopping.py [hunk 9]
```diff
@@ -265,25 +265,28 @@ def validation_epoch_end(self, outputs):
...
+        (3, 2, 0, 6),
+        (5, 2, 0, 10),
         # IF both `min_epochs` and `min_steps` are provided and higher than the `trainer.global_step` when
         # `early_stopping` is being triggered, THEN the highest between `min_epochs * len(train_dataloader)` and
         # `min_steps` would be reached
-        (1, 10),
-        (3, 10),
+        (3, 1, 10, 10),
+        (5, 1, 10, 10),
+        (3, 3, 10, 10),
```
Comment (awaelchli) [inline tests/tests_pytorch/callbacks/test_early_stopping.py:285]: great! I think this is the case for the current PR. The previous stop_step would have been 12, because the check would run only after epoch 4

**Review by awaelchli — APPROVED** [review]

File: tests/tests_pytorch/loops/epoch/test_training_epoch_loop.py [hunk 10]
```diff
@@ -265,3 +266,33 @@ def test_no_val_on_train_epoch_loop_restart(tmpdir):
...
+        (4, None, 1, 4, False, True),
+        (4, 2, 1, 4, False, True),
+        (4, None, 1, 10, True, False),
+        (4, 3, 1, 3, False, True),
+        (4, 10, 1, 10, True, False),
+        (None, 4, 1, 4, False, False),
+        (None, 10, 1, 10, True, False),
+    ],
+)
+def test_should_stop_early_stopping_conditions(
```
Comment (carmocca) [inline tests/tests_pytorch/loops/epoch/test_training_epoch_loop.py:284]: What is the difference to the test in the other file?

Comment (rohitgr7) [reply tests/tests_pytorch/loops/epoch/test_training_epoch_loop.py:284]: the message and conditions of when they would come up respectively.
This one is about epoch loop, which says that the min_values are not met, and training will continue.

the other one stops the training since conditions are met.

Comment (rohitgr7) [reply tests/tests_pytorch/loops/epoch/test_training_epoch_loop.py:284]: let me add descriptions and update test name to make it more clear

File: docs/source-pytorch/common/trainer.rst [hunk 11]
```diff
@@ -1745,3 +1745,57 @@ execution within that function, and the status of the Trainer.
...
+
+If you want to terminate the training during ``.fit``, you can set ``trainer.should_stop=True`` and Lightning will terminate the training
+as soon as possible. Note that, it will respect the arguments ``min_steps`` and ``min_epochs`` to check whether to stop it or not. If these
+arguments are set and the ``current_epoch`` or ``global_step`` doesn't meet these minimum conditions, training will continue until
+both conditions are met. If any of these arguments is not set, it won't be considered for the final decision.
+
+
+.. code-block:: python
+
+    # setting `trainer.should_stop` at any point of training will terminate
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply docs/source-pytorch/common/trainer.rst:1760]: It shows the basic case

File: src/pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 12]
```diff
@@ -102,7 +102,21 @@ def _is_validation_done(self) -> bool:
...
     def done(self) -> bool:
         """Evaluates when to leave the loop."""
-        return (self._is_training_done and self._is_validation_done) or self.trainer.should_stop
+        if self._is_training_done and self._is_validation_done:
+            return True
+
+        if self.trainer.should_stop:
+            # early stopping
+            min_epochs = self.trainer.fit_loop.min_epochs
+            should_stop_early = self.trainer.fit_loop._should_stop_early
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply src/pytorch_lightning/loops/epoch/training_epoch_loop.py:111]: I can move this to `training_epoch_loop`and use `Trainer.min_epochs` to get that property since it is required but then it will use `fit_loop` indirectly in the epoch loop.

File: src/pytorch_lightning/loops/fit_loop.py [hunk 13]
```diff
@@ -170,20 +176,10 @@ def done(self) -> bool:
...
-                rank_zero_debug("`Trainer.fit` stopped: `trainer.should_stop` was set.")
-                return True
-            else:
-                rank_zero_info(
-                    f"Trainer was signaled to stop but the required `min_epochs={self.min_epochs!r}` or"
-                    f" `min_steps={self.min_steps!r}` has not been met. Training will continue..."
-                )
-        self.trainer.should_stop = False
+        if self.trainer.should_stop and self._should_stop_early:
+            rank_zero_debug("`Trainer.fit` stopped: `trainer.should_stop` was set.")
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply src/pytorch_lightning/loops/fit_loop.py:180]: > Should show the same message here?

I don't think so because they have different meanings. That tells the training will continue and this one tells that training will terminate.

> Is there an edge case where both this one and the one in training epoch loop would appear?

no because when the message in training epoch loop occurs like this:
```
if not limit_reached:
    show info message and continue training epoch loop.
```
so it will not come back to fit loop.

File: src/pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (rohitgr7) [reply src/pytorch_lightning/loops/epoch/training_epoch_loop.py:111]: can't find a way, both rely on each other for some property.

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]
